electrumsv/network_support/api_server.py

```diff
@@ -104,6 +114,72 @@ class CapabilitySupport:
+        - prioritisation of mAPI servers
+        - creating a new peer channel using the best available ESVReferenceServer
+        - ensuring there is a record of attempted broadcast & success or failure in
+        MAPIBroadcastCallbacks.
+        - finally, broadcasting through mAPI
+    """
+    server_entries = get_mapi_servers(network, account)
+    if len(server_entries) != 0:
+        await poll_servers_async(server_entries)
+
+    selection_candidates = network.get_api_servers_for_account(account,
+        NetworkServerType.MERCHANT_API)
+
+    broadcast_servers: list[BroadcastCandidate] = prioritise_broadcast_servers(
+        TransactionSize(tx.size()), selection_candidates)
+    if len(broadcast_servers) == 0:
+        raise ServiceUnavailableException("There are no suitable broadcast servers available")
+
+    # Select the best ranked broadcast server
+    broadcast_server = broadcast_servers[0]
+
+    # For the peer channel callbacks to work on public networks we will
+    # require at least one public ESV Reference Server instance for each network
+    peer_channel: Optional[PeerChannel] = None
+    if app_state.config.get('regtest'):
+        selection_candidates = network.get_api_servers_for_account(account,
+            NetworkServerType.GENERAL)
+
+        esv_reference_servers = select_servers(ServerCapability.PEER_CHANNELS, selection_candidates)
+        esv_reference_server = esv_reference_servers[0]  # Todo - prioritise properly
+        aiohttp_client = await network.get_aiohttp_session()
+        assert esv_reference_server.api_server is not None
+        url = esv_reference_server.api_server.url
+        esv_client = ESVClient(url, aiohttp_client, REGTEST_MASTER_TOKEN)
+        peer_channel = await esv_client.create_peer_channel()
+    try:
+        server_id = broadcast_server.candidate.api_server.config.get('id')
+        mapi_callback_row = MAPIBroadcastCallbackRow(
+            tx_hash=tx.hash(),
+            peer_channel_id=peer_channel.channel_id,
+            broadcast_date=datetime.datetime.utcnow().isoformat(),
+            encrypted_private_key=os.urandom(64),  # libsodium encryption not implemented yet
+            server_id=server_id,
+            status_flags=MapiBroadcastStatusFlags.ATTEMPTING
+        )
+        account._wallet.create_mapi_broadcast_callbacks([mapi_callback_row])
+        api_server = broadcast_server.candidate.api_server
+        credential_id = broadcast_server.candidate.credential_id
+        assert api_server is not None
+        return await broadcast_transaction_mapi_simple(tx,
+            api_server, credential_id, peer_channel, merkle_proof, ds_check)
+    except aiohttp.ClientError as e:
+        account._wallet.delete_mapi_broadcast_callbacks(tx_hashes=[tx.hash()])
+        logger.error(f"Error broadcasting to mAPI for tx: {tx.txid()}. Error: {str(e)}")
+        raise
+    finally:
+        updates = [(MapiBroadcastStatusFlags.SUCCEEDED, tx.hash())]
```

Comment:
This happens whether there was an exception or not. It cannot possibly be correct to do this.

Reply:
Yes. You are right. Thanks